lib/compiler-singlepass/src/codegen_x64.rs

```diff
@@ -276,21 +278,208 @@ impl<'a> FuncGen<'a> {
+            } else {
+                break;
+            }
+        }
+
+        if let Some(intrinsic) = self.check_intrinsic(function_index) {
+            self.emit_intrinsic(intrinsic, &params)?
+        } else {
+            let reloc_at = self.assembler.get_offset().0 + self.assembler.arch_mov64_imm_offset();
+            // Imported functions are called through trampolines placed as custom sections.
+            let reloc_target = if function_index < self.module.num_imported_functions {
+                RelocationTarget::CustomSection(SectionIndex::new(function_index))
+            } else {
+                RelocationTarget::LocalFunc(LocalFunctionIndex::new(
+                    function_index - self.module.num_imported_functions,
+                ))
+            };
+            self.relocations.push(Relocation {
+                kind: RelocationKind::Abs8,
+                reloc_target,
+                offset: reloc_at as u32,
+                addend: 0,
+            });
+
+            // RAX is preserved on entry to `emit_call_sysv` callback.
+            // The Imm64 value is relocated by the JIT linker.
+            self.assembler.emit_mov(
+                Size::S64,
+                Location::Imm64(std::u64::MAX),
+                Location::GPR(GPR::RAX),
+            );
+
+            self.emit_call_sysv(
+                |this| {
+                    this.assembler.emit_call_location(Location::GPR(GPR::RAX));
+                },
+                params.iter().copied(),
+            )?;
+
+            self.machine
+                .release_locations_only_stack(&mut self.assembler, &params);
+
+            if !return_types.is_empty() {
+                let ret = self.machine.acquire_locations(
+                    &mut self.assembler,
+                    &[(
+                        return_types[0],
+                        MachineValue::WasmStack(self.value_stack.len()),
+                    )],
+                    false,
+                )[0];
+                self.value_stack.push(ret);
+                if return_types[0].is_float() {
+                    self.assembler
+                        .emit_mov(Size::S64, Location::XMM(XMM::XMM0), ret);
+                    self.fp_stack
+                        .push(FloatValue::new(self.value_stack.len() - 1));
+                } else {
+                    self.assembler
+                        .emit_mov(Size::S64, Location::GPR(GPR::RAX), ret);
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    fn check_intrinsic(&mut self, index: usize) -> Option<Intrinsic> {
+        let function_index = FunctionIndex::new(index);
+        if !self.module.is_imported_function(function_index) {
+            return None;
+        }
+        let signature_index = self.module.functions[function_index];
+        let signature = &self.module.signatures[signature_index];
+        // TODO: maybe speed up this lookup.
+        for import in self.module.imports.keys() {
```

Comment:
Hm, I think we actually must do this. We only check the set of imports on instantiation, not on compilation. So, if one sends a contract with 10_000 imported functions, this will be quite slow.